### ISSUE TITLE
Using apply_feature_transform

### ISSUE THREAD

**Issue by CarlArblue** [lululxvi/deepxde#396]
Hello Mr. Lulu

If I want to replace the input with an output from a model. Is it possible to do it using apply_feature_transform.

I currently have this code:

def input_transform(input):
    x = input[:,0:1]
    # z = input[:,1:2]
    sigma = input[:,2:3]
    t = input[:,3:4]

    x_t = concatenate([x, t], axis=1)
    zeta_output = zeta_call(x_t) # Call our custom zeta_layer

    z = sigma*(zeta_output+hd) # Replace our input z with this output
    z = tf.stop_gradient(z) # Stop the flow of gradients as to not backpropagate through zeta
    
    # FINAL INPUT #
    final_input = concatenate([x, z, sigma, t], axis=1)
    
    return final_input

Will this correctly replace z with the values from my zeta layer? I have tried doing it this way but there seems to be an issue. The gradients with regard to z seems to all be zero. Perhaps I have misunderstood how to use this function. Below is the apply_output_transform I am using. Just in case there was no issue with my feature transform.

def output_transform(input, outputs):
    x = input[:,0:1]                                     
    z = input[:,1:2]                             
    # sigma = input[:,2:3]
    t = input[:,3:4]

    x_t = concatenate([x, t], axis=1)
    x_z_t = concatenate([x, z, t], axis=1)  

    zeta_output = zeta_call(x_t)
    phi_output = phi_call(x_z_t)

    # FINAL OUTPUT #
    final_output = concatenate([zeta_output, phi_output], axis=1)

    return final_output

On a side note. Is it possible to included a tf.gradients as a model output. Something akin to:

def output_transform(input, outputs):
    x = input[:,0:1]                                     
    z = input[:,1:2]                             
    sigma = input[:,2:3]
    t = input[:,3:4]

    x_t = concatenate([x, t], axis=1)
    x_z_t = concatenate([x, z, t], axis=1)  

    zeta_output = zeta_call(x_t)
    phi_output = phi_call(x_z_t)

    z_sigma = sigma*(zeta_output+hd)
    z_sigma = tf.stop_gradient(z_sigma)
    x_zsigma_t = concatenate([x, z_sigma, t], axis=1)  
    phi_tilde_output = phi_call(x_zsigma_t)
    w_sigma = tf.gradients(phi_tilde_output, z_sigma)

    # FINAL OUTPUT #
    final_output = concatenate([zeta_output, phi_output, w_sigma], axis=1)

    return final_output

I am looking forward to hearing from you.

Thank you

**Comment by lululxvi** [comment]
Your input transform looks good, but why you need `tf.stop_gradient(z)`?

It is OK to add tf.gradients as output.

**Comment by CarlArblue** [comment]
Hi Mr. Lu lu,

So I kept looking around and using TensorFlows description of tf.gradients(https://www.tensorflow.org/api_docs/python/tf/gradients) it states that if you want partial derivatives instead of total derivatives you should use tf.stop_gradient. So if I want the partial derivative wrt. zeta(x,t) I should use stop_gradients. Do you know anymore about this or if it would cause issues?

Thank you for your time.

**Comment by lululxvi** [comment]
Not familiar with tf.stop_gradient

**Comment by CarlArblue** [comment]
Hello again.

So I have been trying to use apply_transform with PyTorch. I can get the model to compile but the loss does not decrease. I believe there is an error in how I have create the sequential model in PyTorch. Perhaps you are able to tell me where my error is? I have supplied the code below:

# ZETA LAYERS
zeta_layer_size = 15
pytorch_zeta_call = torch.nn.Sequential(
    torch.nn.Flatten(),
    torch.nn.LazyLinear(out_features=zeta_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=zeta_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=zeta_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=1, bias=False)
    )

# PHI LAYERS
phi_layer_size = 20
pytorch_phi_call = torch.nn.Sequential(
    torch.nn.Flatten(),
    torch.nn.LazyLinear(out_features=phi_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=phi_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=phi_layer_size, bias=False),
    torch.nn.Tanh(),
    torch.nn.LazyLinear(out_features=1, bias=False)
    )

def output_transform(input, outputs):
    x = input[:,0:1]
    z = input[:,1:2]
    t = input[:,2:3]

    x_t = torch.concat([x,t], axis=1)
    x_z_t = torch.concat([x, z, t], axis=1)

    zeta_output = pytorch_zeta_call(x_t)
    phi_output = pytorch_phi_call(x_z_t)

    # FINAL OUTPUT #
    final_output = torch.concat([zeta_output, phi_output], axis=1)

    return final_output

def input_transform(input):
    x = input[:,0:1]
    sigma = input[:,1:2]
    t = input[:,2:3]

    x_t = torch.concat([x,t], axis=1)

    zeta = pytorch_zeta_call(x_t)

    z = sigma*(zeta+hd)-hd
    # z = tf.stop_gradient(z)
    
    # FINAL INPUT #
    final_input = torch.concat([x, z, t], axis=1)
    
    return final_input

FNN_layer = [3,2] # x, sigma, t -> zeta, phi
net = dde.nn.FNN(FNN_layer, "tanh", "Glorot uniform")
net.apply_feature_transform(input_transform)
net.apply_output_transform(output_transform)
model = dde.Model(data, net)

Thanks in advance.

**Comment by lululxvi** [comment]
See example: https://github.com/lu-group/sbinn
